### PR TITLE
[Serializer] rename exception interface

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by unkind** [comment]
This empty interface looks oddly. Maybe it makes sense to add base class of Serializer exceptions for 3.0 and remove interface completely? 

I think extending base exceptions like `\UnexpectedValueException`, `\InvalidArgumentException`, etc. is pointless: I'd never try to catch them if I know package specific exception like `\Symfony\Component\Serializer\Exception\InvalidArgumentException`.

**Comment by Tobion** [comment]
@unkind the SPL exception give semantic meaning (logic vs runtime). so they are not useless IMO.

**Comment by unkind** [comment]
I don't say they are useless, I don't like idea of having this structure:
`\Symfony\Component\Serializer\Exception\InvalidArgumentException extends \InvalidArgumentException`
`\Symfony\Component\Serializer\Exception\UnexpectedValueException extends \UnexpectedValueException`
etc.
What's the point?

**Comment by nicolas-grekas** [comment]
The point is domain separation - each domain should have its own domain-specific exceptions, so that you can layer domains and still know which layer is at fault. That's also the reason for wrapping exceptions in domain specific exceptions (thus ->getPrevious() on the base Exception class)

**Comment by unkind** [comment]
> The point is domain separation - each domain should have its own domain-specific exceptions

Right, but why do you need SPL exceptions here? Any case when you would try to catch generic `\UnexpectedValueException`? I suggest to inherit something like `SerializerException` instead in order to allow to catch all serializer exceptions (I guess empty exception interface has the same purpose).

My point is that empty interface indicates a problem. I know concept of marker interface, but it's different case IMO.

**Comment by dunglas** [comment]
:+1: for this change.

**Comment by dunglas** [comment]
As explained by @nicolas-grekas, use cases are:
- being to able to catch only exception thrown by the Serializer Component
- being able to catch all logic or runtime exceptions.

Think about a generic Logger, not aware of libraries used in the project, and logging all `\RuntimeException`. Thanks to `\RuntimeException`, this logger can catch both `Assetic\Exception\ FilterException` and `Symfony\Component\Serializer\Exception\RuntimeException`.

If `Symfony\Component\Serializer\Exception\RuntimeException` doesn't extend  `\RuntimeException` it's not possible.

Btw, this is done the same way other Symfony components: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Validator/Exception/RuntimeException.php

**Comment by unkind** [comment]
> being able to catch all logic or runtime exceptions thrown by different libraries

I can understand catching just `\RuntimeException`, because it's a little bit special one: like in Java, I usually use it for exceptions which is not documented in `@throws` annotation, i.e. some programmer's mistakes, it's not API's one. 

But seriously, I don't get any case for `\InvalidArgumentException` and `\UnexpectedValueException`. They tell me literally nothing about context: somewhere we've passed invalid value... OK, so what?

**Comment by dunglas** [comment]
@unkind To be able to catch `\RuntimeException`, we need the `Symfony\Component\Serializer\Exception\ExceptionInterface` and we need to extend SPL exceptions. So it's better leverage the SPL and extends the more precise exception such as `\InvalidArgumentException`. Even if there are fewer use cases for them, what's the point with not extending it but extends directly `\RuntimeException`? For instance `\UnexpectedValueException` extends `\RuntimeException`.

**Comment by unkind** [comment]
OK, I have to clarify my opinion: I don't like idea of inheriting `\RuntimeException` and `\LogicException` at all: they are "bad" exceptions, programmer's mistakes, like misconfigured Serializer. Usually, at application level you can't "fix" it, program should fail with error. So you are able to omit them in the `@throws` annotation. If you are pretty sure that you don't want to fail, you can just catch `\Exception`:
```php
try {
    return $serializer->serialize($foo);
} catch (\Exception $e) {
    $this->logger->error('We are screwed and I have no idea why', ['exception' => $e]);

    return '';
}
```
but usually I'd prefer to fail by default, because ignoring error can lead serious bugs in the future (http://en.wikipedia.org/wiki/Fail-fast).

For example, `Serializer::normalizeObject` throws an `UnexpectedValueException`, but it's a private method. Who catches it on Serializer component level? No one. Who declares it in the `@throws` annotation in some public interface of Serializer component? No one. So to be consistent you have to either put `@throws UnexpectedValueException` in the `NormalizerInterface::normalize` or remove runtime/logic exceptions from public contracts at all. I like the last one.

But for bad incoming data you have to throw regular (non-runtime) exception which would extend `SerializerException`.

**Comment by nicolas-grekas** [comment]
I'm not sure I understand where you want to go @unkind. Anyway, this discussion is not related to this PR. Here I only want to fix a class that does not respect a naming convention applied everywhere otherwise (Interface suffix for interfaces).
Feel free to open an issue, or better, a PR, to expose your arguments.

File: src/Symfony/Component/Serializer/Exception/ExceptionInterface.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+namespace Symfony\Component\Serializer\Exception;
+
+/**
+ * Base exception interface.
+ *
+ * @author Johannes M. Schmitt <schmittjoh@gmail.com>
+ */
+interface ExceptionInterface extends Exception
```
Comment (Taluu) [inline src/Symfony/Component/Serializer/Exception/ExceptionInterface.php]: Shouldn't be the other way around ? `Exception` extending the `ExceptionInterface` ?

Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Exception/ExceptionInterface.php]: Nope: this way allows the best forward/backward compatibility

Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Exception/ExceptionInterface.php]: Nope: this way allows the best forward/backward compatibility

**Comment by fabpot** [comment]
:+1:

**Comment by fabpot** [comment]
Before merging, @nicolas-grekas Can you add a note in the CHANGELOG of the component and a note in the UPGRADE file for 3.0?

**Comment by nicolas-grekas** [comment]
Notes added (is it enough?)

**Comment by fabpot** [comment]
Looks good to me.

**Comment by Tobion** [comment]
:+1:

**Comment by dunglas** [comment]
Thank you @nicolas-grekas.

**Comment by dunglas** [comment]
Thank you @nicolas-grekas.

**Pull request merged**